atomic_reactor/odcs_util.py

```diff
@@ -0,0 +1,120 @@
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+
+from atomic_reactor.util import get_retrying_requests_session
+
+import logging
+import time
+
+
+logger = logging.getLogger(__name__)
+
+
+OIDC_TOKEN_HEADER = 'OIDC_access_token'
+OIDC_CLAIM_SCOPE_HEADER = 'OIDC_CLAIM_scope'
+
+
+class ODCSClient(object):
+    def __init__(self, url, insecure=False, token=None):
+        if url.endswith('/'):
+            self.url = url
+        else:
+            self.url = url + '/'
+        self.insecure = insecure
+        self.token = token
+        # method_whitelist=False allows retrying non-idempotent methods like POST
+        self.session = get_retrying_requests_session(method_whitelist=False)
+
+    def _auth_headers(self):
+        headers = {}
+        if self.token:
+            headers[OIDC_TOKEN_HEADER] = self.token
+            headers[OIDC_CLAIM_SCOPE_HEADER] = ('openid https://id.fedoraproject.org/scope/groups '
+                                                'https://pagure.io/odcs/new-compose '
+                                                'https://pagure.io/odcs/renew-compose '
+                                                'https://pagure.io/odcs/delete-compose')
+            # FIXME: passing the claim scope in the request doesn't make any sense,
+            #   because we're not to be trusted for what scopes we've claimed, but
+            #   match what the server wants.
+
+        return headers
+
+    def start_compose(self, source_type, source, packages=None, sigkeys=None):
+        body = {
+            'source': {
+                'type': source_type,
+                'source': source
+            }
+        }
+        if packages:
+            body['source']['packages'] = packages
+
+        if sigkeys is not None:
+            body['source']['sigkeys'] = sigkeys
+
+        logger.info("Starting compose for source_type={source_type}, source={source}"
+                    .format(source_type=source_type, source=source))
+        response = self.session.post(self.url + 'composes/',
+                                     json=body,
+                                     headers=self._auth_headers())
+        response.raise_for_status()
+
+        return response.json()
+
+    def renew_compose(self, compose_id):
+        """Renew, or extend, existing compose
+
+        If the compose has already been removed, ODCS creates a new compose.
+        Otherwise, it extends the time_to_expire of existing compose. In most
+        cases, caller should assume the compose ID will change.
+        """
+        logger.info("Renewing compose %d", compose_id)
+        response = self.session.patch(self.url + 'composes/' + str(compose_id),
+                                      headers=self._auth_headers())
+        response.raise_for_status()
+        response_json = response.json()
+        compose_id = response_json['id']
```

Comment:
From the docs it doesn't seem to be the case. The `raise_for_status` above should raise an exception if something went wrong with the compose request.

I think if it got here and there's no `id` filed in the response json, it's definitely unexpected behavior. Fetching `id` like this would make such scenario obvious.

I'm inclined to think no changes here. What do you think?

Reply:
Agreed.